File: year2020/day05/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return max(input)


def part_2(input):
  input.sort()
  return next((seat - 1 for seat, prev_seat in zip(input[1:], input) if seat - prev_seat != 1), 0)


def setup(path):
  global input

  with open(path, 'r') as file:
    input = setup_seats([[line[:7], line[-3:]] for line in file.read().split('\n')])


def setup_seats(input):
  seats = []
  for pass_ in input:
    rl, ru, cl, cu = 0, 127, 0, 7
    for char in pass_[0]:
      if char == 'F':
        ru = (rl + ru) // 2
      else:
        rl = (rl + ru) // 2 + 1

    for char in pass_[1]:
      if char == 'L':
        cu = (cl + cu) // 2
      else:
        cl = (cl + cu) // 2 + 1

    seats.append(rl * 8 + cl)
  return seats


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()

File: year2020/day05/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: string[][] = file.toString().split('\n').map((pass) => [pass.slice(0, 7), pass.slice(-3)]);
const seats: number[] = [];
setupSeats();
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return Math.max(...seats)
}

export function part_2(): number {
  return seats.sort((a, b) => a - b).find((seat, index) => index > 0 && seat - seats[index - 1] !== 1) - 1 || 0;
}

function setupSeats(): void {
  input.forEach((pass) => {
    let [rl, ru, cl, cu] = [0, 127, 0, 7];
    pass[0].split('').forEach((char) => char === 'F' ? ru = Math.floor((rl + ru) / 2) : rl = Math.ceil((rl + ru) / 2));
    pass[1].split('').forEach((char) => char === 'L' ? cu = Math.floor((cl + cu) / 2) : cl = Math.ceil((cl + cu) / 2));
    seats.push(rl * 8 + cl);
  });
}